### PR TITLE
App code M3

### PR DESCRIPTION
**Pull request by android-dev-lxl**
- Added app code
- UX design review feedback
- Addressed team feedback

### CODE REVIEW

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 1]
```diff
@@ -37,8 +67,132 @@ class MainActivity : ComponentActivity() {
...
                     color = MaterialTheme.colorScheme.background
                 ) {
+                    TipTimeScreen()
                 }
             }
         }
     }
 }
+
+@Composable
```
Comment (kkuan2011) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:77]: Will you also be able to include the changes from the design review? (e.g. top app bar, background color, etc..)

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 2]
```diff
@@ -37,8 +67,132 @@ class MainActivity : ComponentActivity() {
...
+    var roundUp by remember { mutableStateOf(false) }
+
+    val amount = amountInput.toDoubleOrNull() ?: 0.0
+    val tipPercent = tipInput.toDoubleOrNull() ?: 0.0
+    val tip = calculateTip(amount, tipPercent, roundUp)
+
+    val focusManager = LocalFocusManager.current
+
+    Column(
+        modifier = Modifier.padding(32.dp),
```
[the remark this answers is not in the document]
Comment (android-dev-lxl) [reply app/src/main/java/com/example/tiptime/MainActivity.kt:90]: We already use vertical arrangement, so I added horizontal arrangement. Would that be confusing? 

```
Column(
        modifier = modifier.padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) 
```

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -29,7 +29,6 @@
         <activity
             android:name=".MainActivity"
             android:exported="true"
-            android:label="@string/app_name"
             android:theme="@style/Theme.TipTime">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
```
Comment (johnshea) [inline app/src/main/AndroidManifest.xml:34]: Remove the following blank line.

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 4]
```diff
@@ -37,8 +67,149 @@ class MainActivity : ComponentActivity() {
...
+                .wrapContentWidth(Alignment.End),
+            checked = roundUp,
+            onCheckedChange = onRoundUpChanged,
+        )
+    }
+}
+
+/**
+ * Calculates the tip based on the user input and format the tip amount
+ * according to the local currency and display it onscreen.
```
Comment (johnshea) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:199]: nit: This function does not "display it onscreen" as the comment states.

File: app/src/main/java/com/example/tiptime/ui/theme/Theme.kt [hunk 5]
```diff
@@ -30,23 +30,77 @@ import androidx.compose.ui.platform.LocalContext
 import androidx.compose.ui.platform.LocalView
 import androidx.core.view.WindowCompat
 
-private val DarkColorScheme = darkColorScheme(
-    primary = Purple80,
-    secondary = PurpleGrey80,
-    tertiary = Pink80
-)
```
Comment (johnshea) [inline app/src/main/java/com/example/tiptime/ui/theme/Theme.kt:33]: Remove blank line

File: app/src/main/res/drawable/percent.xml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<vector android:height="24dp" android:tint="#000000"
```
Comment (johnshea) [inline app/src/main/res/drawable/percent.xml:1]: Does this need a license header?

**Review by johnshea — CHANGES_REQUESTED** [review]
Nit: Inconsistent usage of "http" / "https" in the copyright headers.

File: app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (johnshea) [inline app/src/main/AndroidManifest.xml:34]: Remove the following blank line.

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 7]
```diff
@@ -37,8 +67,149 @@ class MainActivity : ComponentActivity() {
...
+) {
+    Row(
+        modifier = modifier
+            .fillMaxWidth()
+            .size(48.dp),
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        Text(text = stringResource(R.string.round_up_tip))
+        Switch(
+            modifier = modifier
```
[the remark this answers is not in the document]
Comment (android-dev-lxl) [reply app/src/main/java/com/example/tiptime/MainActivity.kt:188]: We need to use the passed in Modifier. I fixed the issue .

[the remark this answers is not in the document]
Comment (johnshea) [reply app/src/main/java/com/example/tiptime/MainActivity.kt:189]: You did use it for the `Row` composable function.

[the remark this answers is not in the document]
Comment (android-dev-lxl) [reply app/src/main/java/com/example/tiptime/MainActivity.kt:189]: Removing it fixed it 
image

**Review by yenerm — COMMENTED** [review]
lgtm

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 8]
```diff
@@ -37,8 +70,143 @@ class MainActivity : ComponentActivity() {
...
                 ) {
+                    TipTimeLayout()
                 }
             }
         }
     }
 }
+
+@Composable
+fun TipTimeLayout(modifier: Modifier = Modifier) {
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:81]: don't need a modifier on the main Composable

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 9]
```diff
@@ -37,8 +70,143 @@ class MainActivity : ComponentActivity() {
...
+            keyboardOptions = KeyboardOptions.Default.copy(
+                keyboardType = KeyboardType.Number,
+                imeAction = ImeAction.Next
+            ),
+            keyboardActions = KeyboardActions(
+                onNext = { focusManager.moveFocus(FocusDirection.Down) }
+            ),
+            value = amountInput,
+            onValueChanged = { amountInput = it },
+            modifier = Modifier.padding(bottom = 32.dp)
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:118]: 32.dp is repeated over 3x - let's add it to the dimens file - same with any repeated dimen values

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 10]
```diff
@@ -37,8 +70,143 @@ class MainActivity : ComponentActivity() {
...
+    value: String,
+    onValueChanged: (String) -> Unit,
+    modifier: Modifier = Modifier
+) {
+    TextField(
+        value = value,
+        singleLine = true,
+        leadingIcon = { Icon(painter = painterResource(id = leadingIcon), null) },
+        modifier = modifier.fillMaxWidth(),
+        shape = RoundedCornerShape(topStart = 16.dp, topEnd = 16.dp),
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:163]: add 16.dp to dimens

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 11]
```diff
@@ -37,8 +71,145 @@ class MainActivity : ComponentActivity() {
...
+    onValueChanged: (String) -> Unit,
+    modifier: Modifier = Modifier
+) {
+    val mediumPadding = dimensionResource(R.dimen.padding_medium)
+    TextField(
+        value = value,
+        singleLine = true,
+        leadingIcon = { Icon(painter = painterResource(id = leadingIcon), null) },
+        modifier = modifier.fillMaxWidth(),
+        shape = RoundedCornerShape(topStart = mediumPadding, topEnd = mediumPadding),
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:166]: You could add this to Shapes.kt file

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 12]
```diff
@@ -37,8 +71,145 @@ class MainActivity : ComponentActivity() {
...
+    modifier: Modifier = Modifier
+) {
+    Row(
+        modifier = modifier
+            .fillMaxWidth()
+            .size(48.dp),
+    ) {
+        Text(text = stringResource(R.string.round_up_tip))
+        Switch(
+            modifier = modifier
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:187]: modifier = Modifier

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 13]
```diff
@@ -37,8 +71,145 @@ class MainActivity : ComponentActivity() {
...
+    value: String,
+    onValueChanged: (String) -> Unit,
+    modifier: Modifier = Modifier
+) {
+    val mediumPadding = dimensionResource(R.dimen.padding_medium)
+    TextField(
+        value = value,
+        singleLine = true,
+        leadingIcon = { Icon(painter = painterResource(id = leadingIcon), null) },
+        modifier = modifier.fillMaxWidth(),
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:165]: pass this in through the parent

File: app/src/main/java/com/example/tiptime/MainActivity.kt [hunk 14]
```diff
@@ -37,8 +71,145 @@ class MainActivity : ComponentActivity() {
...
+        )
+        RoundTheTipRow(
+            roundUp = roundUp,
+            onRoundUpChanged = { roundUp = it },
+            modifier = Modifier.padding(bottom = largePadding)
+        )
+        Text(
+            text = stringResource(R.string.tip_amount, tip),
+            style = MaterialTheme.typography.displaySmall,
+            fontWeight = FontWeight.Bold
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/tiptime/MainActivity.kt:144]: font weight can be included in Type.kt as part of the displaySmall parameters

File: app/src/main/java/com/example/tiptime/MainActivity.kt (hunk 11 above)
Comment (android-dev-lxl) [reply app/src/main/java/com/example/tiptime/MainActivity.kt:166]: This is just one shape and this is a fairly complex codelab in the earlier units. 
Could we just leave it as is?

Comment (MagicalMeghan) [reply app/src/main/java/com/example/tiptime/MainActivity.kt:165]: ok

**Review by MagicalMeghan — APPROVED** [review]

**Pull request merged**